### PR TITLE
BASW-733: Creating first version of data health checker extension

### PR DESCRIPTION
**Pull request by omarabuhussein**
# Overview

This extension performs regular data health checks on Membershipextras and CiviCRM financial data. The extension defines a new scheduled job called : "Membershipextras - data health checker":

2021-06-22 01_46_23-Microsoft Store

that runs daily and update the database with the health check results.

To access the results you can go to "Administer >> CiviContribute >> Data Health Checker Dashboard", which will show different type of issues in the existing data categorized by their type along with the number of the affected records, from there you can export any category records.

2021-06-22 01_52_38-ME Health Checker Issues Dashboard _ civi5352s1

Currently, we are checking for the following type of errors:

- Contributions paid with Direct Debit but with no related Payment Plan: Direct debit payments are mostly paid with Payment Plans, so if a
  direct debit contribution does not have a linked payment-plan then probably there is an issue in that contribution.
- Missing financial transaction records on contributions: Contributions should at least have one financial transaction record and one entity
  financial transaction record, but if the contribution have non, then there is an issue with it.
- Missing line items on contributions: Any contribution should at least have one line item.
- Missing financial items on line items: Line items should at least have one financial item and entity financial transaction record.
- Contributions with mismatched line items amount: Contribution amount should equal the sum of all its line items amounts plus taxes.
- Contributions with mismatched line items tax amount: Contribution tax amount should equal the sum of all its line items tax amounts.
- Offline payment plans but with no related memberships: So far we only support payment plans with memberships, if there is a payment plan but it does not
  have a linked membership then it means there is an issue with it.
- Offline payment plans but with auto-renew flag set to False: All payment plans should be renewable
- Offline payment plans but without any subscription line items: Each payment plan should at least have one subscription line item
- Direct debit payment plans but with no related mandate: Payment plans paid with direct debit should have a mandate linked to them.
- Direct debit contributions but with no related mandate: Contributions paid with direct debit should have a mandate linked to them.

### CODE REVIEW

File: medatahealthchecker.php [hunk 1]
```diff
@@ -142,3 +142,19 @@ function medatahealthchecker_civicrm_entityTypes(&$entityTypes) {
...
+
+/**
+ * Implements hook_civicrm_navigationMenu().
+ */
+function medatahealthchecker_civicrm_navigationMenu(&$menu) {
+  $issuesDashboardMenuItem = [
+    'name' => 'medatahealthchecker_issues_dashboard',
+    'label' => ts('Data Health Checker Dashboard'),
+    'url' => 'civicrm/medatahealthchecker/issues-dashboard',
+    'permission' => 'administer CiviCRM,administer MembershipExtras',
```
Comment (ahed-compucorp) [inline medatahealthchecker.php:154]: Can you double check the permission `administer MembershipExtras`?
I cannot see it mentioned here or in `uk.co.compucorp.membershipextras` extension's docs/readme

**Review by ahed-compucorp — CHANGES_REQUESTED** [review]
These are important step and I see this extensions will help us a lot in the long run, I have the following notes:
- Can you add a workflow action to do the lint checks?
- Can you show more love for phpDoc? :crossed_fingers: 
- I don't see a need for the Log suffix in the name of `IssuesLog` entity, maybe going with `Issue` is  a good one.

File: medatahealthchecker.php (hunk 1 above)
Comment (omarabuhussein) [reply medatahealthchecker.php:154]: > Can you add a workflow action to do the lint checks?

will do

> Can you show more love for phpDoc? 🤞

Unless you think certain classes and/or methods are not clear and could benefit from it and that you can point them out, I don't think there is a need. 

> I don't see a need for the Log suffix in the name of IssuesLog entity, maybe going with Issue is a good one.

Don't know but I feel like `Issue` alone would be a weird name, do you have other suggestions ? if not I will just keep it as is.

> Can you double check the permission administer MembershipExtras?
I cannot see it mentioned here or in uk.co.compucorp.membershipextras extension's docs/readme

It is defined here: https://github.com/compucorp/uk.co.compucorp.membershipextras/blob/master/membershipextras.php#L477

Comment (ahed-compucorp) [reply medatahealthchecker.php:154]: > Unless you think certain classes and/or methods are not clear

I can see that simply reading the method reveals what it is supposed to do and that is enough.


> Don't know but I feel like Issue

I see.


Proceed with the workflow action.

**Review by ahed-compucorp — APPROVED** [review]

**Pull request merged**